geoopt/manifolds/stereographic/math.py

```diff
@@ -0,0 +1,2018 @@
+        =
+        v \mapsto \lambda^\kappa_p v
+
+
+    Finally, having the points :math:`x,\:y \in \mathcal{M}_\kappa^n` and a
+    tangent vector :math:`u\in T_x\mathcal{M}_\kappa^n` we obtain
+
+    .. math::
+
+        P^\kappa_{x\to y}(v)
+        =
+        (U^\kappa_y)^{-1}\left(\operatorname{gyr}[y, -x] U^\kappa_x(v)\right)\\
+        =
+        \operatorname{gyr}[y, -x] v \lambda^\kappa_x / \lambda^\kappa_y
+
+    .. plot:: plots/extended/stereographic/parallel_transport.py
+
+
+    Parameters
+    ----------
+    x : tensor
+        starting point
+    y : tensor
+        end point
+    v : tensor
+        tangent vector at x to be transported to y
+    k : tensor
+        sectional curvature of manifold
+    dim : int
+        reduction dimension for operations
+
+    Returns
+    -------
+    tensor
+        transported vector
+    """
+    return _parallel_transport(x, y, v, k, dim=dim)
+
+
+@torch.jit.script
+def _parallel_transport(
+    x: torch.Tensor, y: torch.Tensor, u: torch.Tensor, k: torch.Tensor, dim: int = -1
+):
+    return (
+        _gyration(y, -x, u, k, dim=dim)
+        * _lambda_x(x, k, keepdim=True, dim=dim)
+        / _lambda_x(y, k, keepdim=True, dim=dim)
+    )
+
+
+def parallel_transport0(y: torch.Tensor, v: torch.Tensor, *, k: torch.Tensor, dim=-1):
+    r"""
+    Compute the parallel transport of :math:`v` from the origin :math:`0` to :math:`y`.
+
+    This is just a special case of the parallel transport with the starting
+    point at the origin that can be computed more efficiently and more
+    numerically stable.
+
+    Parameters
+    ----------
+    y : tensor
+        target point
+    v : tensor
+        vector to be transported from the origin to y
+    k : tensor
+        sectional curvature of manifold
+    dim : int
+        reduction dimension for operations
+
+    Returns
+    -------
+    tensor
+    """
+    return _parallel_transport0(y, v, k, dim=dim)
+
+
+@torch.jit.script
+def _parallel_transport0(
+    y: torch.Tensor, v: torch.Tensor, k: torch.Tensor, dim: int = -1
+):
+    return v * (1 + k * y.pow(2).sum(dim=dim, keepdim=True)).clamp_min(1e-15)
+
+
+def parallel_transport0back(
+    x: torch.Tensor, v: torch.Tensor, *, k: torch.Tensor, dim: int = -1
+):
+    r"""
+    Perform parallel transport to the zero point.
+
+    Special case parallel transport with last point at zero that
+    can be computed more efficiently and numerically stable
+
+    Parameters
+    ----------
+    x : tensor
+        target point
+    v : tensor
+        vector to be transported
+    k : tensor
+        sectional curvature of manifold
+    dim : int
+        reduction dimension for operations
+
+    Returns
+    -------
+    tensor
+    """
+    return _parallel_transport0back(x, v, k=k, dim=dim)
+
+
+@torch.jit.script
+def _parallel_transport0back(
+    x: torch.Tensor, v: torch.Tensor, k: torch.Tensor, dim: int = -1
+):
+    return v / (1 + k * x.pow(2).sum(dim=dim, keepdim=True)).clamp_min(1e-15)
```

Comment:
Maybe fuse here.

Reply:
is there anything to fuse?

Reply:
`addcmul` instead of `1 + a * b`

Reply:
https://github.com/pytorch/pytorch/issues/23655

Reply:
I see torch script should do this for me, no need to change

Reply:
I'm not sure if it actually fuses ops. Inspecting the resulting graph I do not see any fused ops

Reply:
I always forget that all the ops are serialized now. Though not sure how fusing handled by torchscript. Anyay, resolved.